### PR TITLE
Allow openapi definitions to be built for a single resource

### PR DESCRIPTION
**Pull request by jennybuckley**
This PR allows for the openapi definitions to be built for a single resource. Needed for server side apply.

**Implementation:**

Split the init([]*restful.WebService) function into three functions:
- init(): runs before the swagger spec is built
- buildPaths([]*restful.WebService): The function which actually builds the swagger spec
- finalizeSwagger(): runs after the spec is built

This way we can substitute a different function in for buildPaths, which doesn't need to take a web service as an argument.

cc @mbohlool @apelisse @seans3

### CODE REVIEW

File: pkg/builder/openapi.go [hunk 1]
```diff
@@ -45,7 +45,46 @@ type openAPI struct {
 
 // BuildOpenAPISpec builds OpenAPI spec given a list of webservices (containing routes) and common.Config to customize it.
 func BuildOpenAPISpec(webServices []*restful.WebService, config *common.Config) (*spec.Swagger, error) {
-	o := openAPI{
+	o := newOpenAPI(config)
+
+	err := o.init()
```
Comment (apelisse) [inline pkg/builder/openapi.go]: Nit: It's a private struct so it doesn't matter that much, but you've created an implicit dependency in the order in which these functions need to be called. Can you call `init` in `newOpenAPI`, since you already have a sort of constructor? 

`finishSpec` is also suffering from these syndroms, you need to call it before you can get the `o.swagger`, if you forget to call it then, who knows what will happen. Can we "finishSpec" before (like in init) or do we need all the data before? Then you would have:

```
o, err := newOpenAPIConfig(config)
// handle err
o.addPath(webservices)
return o.Swagger() // But this function should be const.
```

File: pkg/builder/openapi.go [hunk 2]
```diff
@@ -45,7 +45,46 @@ type openAPI struct {
...
+	err = o.finishSpec()
+	if err != nil {
+		return nil, err
+	}
+
+	return o.swagger, nil
+}
+
+// BuildOpenAPISpecForResource builds a partial OpenAPI spec given a sample object and common.Config to customize it.
+func BuildOpenAPISpecForResource(model interface{}, config *common.Config) (*spec.Swagger, error) {
```
Comment (apelisse) [inline pkg/builder/openapi.go]: Maybe you should split into different PRs or at least commits when you're doing different things (refactor AND add new methods)

File: pkg/builder/openapi.go (hunk 1 above)
Comment (jennybuckley) [reply pkg/builder/openapi.go]: Can definitely call init in newOpenAPI, but we do need the spec to be built before calling finishSpec (it contains things like a postProcessSpec function, which is defined by the openapiconfig)

Comment (jennybuckley) [reply pkg/builder/openapi.go]: Would a function like this be any better?
```
func createSpec(config common.OpenAPIConfig, buildSwaggerFunc func(o *openAPI)) *spec.Swagger {
o := openAPI{
...
}
o.init()
buildSwaggerFunc(o)
o.finish()
return o.swagger
}
```
(plus some error handling)

Comment (apelisse) [reply pkg/builder/openapi.go]: ~~Meh, I don't know if you have many options here. It looks like the `openAPI` struct is having too many side-effects already. Let's see how we can make things better without changing everything ...~~

~~I'm tempted to think that these methods in general should be free-functions instead (because they don't really maintain a cohesive state for the struct anyway).~~

Yeah, maybe you should leave it as it is. :-)

Comment (jennybuckley) [reply pkg/builder/openapi.go]: Okay, and I added a comment documenting the requirement to call finish before returning the spec.

**Comment by jennybuckley** [comment]
@seans3 updated to add another unit test, the coverage is now up from 72 to 74%

File: pkg/builder/openapi.go [hunk 3]
```diff
@@ -83,22 +95,25 @@ func (o *openAPI) init(webServices []*restful.WebService) error {
...
-	err := o.buildPaths(webServices)
-	if err != nil {
-		return err
-	}
+	return o
+}
+
+// getSwagger is called after the spec is built and returns the final spec.
+// NOTE: getSwagger also make changes to the final spec, as specified in the config.
+func (o *openAPI) getSwagger() (*spec.Swagger, error) {
```
Comment (seans3) [inline pkg/builder/openapi.go]: Consider the name "finalizeSwagger" to emphasize that the swagger is modified before being returned

File: pkg/builder/openapi.go [hunk 4]
```diff
@@ -45,6 +45,28 @@ type openAPI struct {
 
 // BuildOpenAPISpec builds OpenAPI spec given a list of webservices (containing routes) and common.Config to customize it.
 func BuildOpenAPISpec(webServices []*restful.WebService, config *common.Config) (*spec.Swagger, error) {
+	o := newOpenAPI(config)
+	err := o.buildPaths(webServices)
```
Comment (seans3) [inline pkg/builder/openapi.go]: This looks much cleaner, since the "buildPaths" is not lost in the "init"

File: pkg/builder/openapi.go (hunk 2 above)
Comment (mbohlool) [reply pkg/builder/openapi.go]: It is actually not valid if it does not have paths.

File: pkg/builder/openapi.go (hunk 3 above)
Comment (jennybuckley) [reply pkg/builder/openapi.go]: Done.
Sorry, can you lgtm again? I had to recommit.

File: pkg/builder/openapi.go [hunk 5]
```diff
@@ -45,6 +45,28 @@ type openAPI struct {
...
+	if err != nil {
+		return nil, err
+	}
+	return o.finalizeSwagger()
+}
+
+// BuildOpenAPIDefinitionsForResource builds a partial OpenAPI spec given a sample object and common.Config to customize it.
+func BuildOpenAPIDefinitionsForResource(model interface{}, config *common.Config) (*spec.Swagger, error) {
```
Comment (mbohlool) [inline pkg/builder/openapi.go]: Can we return definitions instead of spec.Swagger here? the return spec is not a valid spec and hence we should not return it in my opinion.

Comment (jennybuckley) [reply pkg/builder/openapi.go]: Okay
